homeassistant/components/airzone_cloud/climate.py

```diff
@@ -278,9 +304,58 @@ def __init__(
         ]
         if HVACMode.OFF not in self._attr_hvac_modes:
             self._attr_hvac_modes += [HVACMode.OFF]
+        if (
+            self.get_airzone_value(AZD_SPEED) is not None
+            and self.get_airzone_value(AZD_SPEEDS) is not None
+        ):
+            self._set_fan_speeds()
 
         self._async_update_attrs()
 
+    def _set_fan_speeds(self) -> None:
```

Comment:
I have the feeling this function is probably a tad too complex, but I have no better way.

But if I understand correctly, we will create fan modes depending on what the fan can do. Is this maybe something to move to the lib? Is this HA specific?

Reply:
This is pretty similar to https://github.com/home-assistant/core/blob/c83388fd2da3fab062dec3fc3c3ebc8013ca478b/homeassistant/components/airzone/climate.py#L152-L171 but as opposed to the local integration, in which all devices seems to support 0 (auto) speed, some older cloud devices can only be set to manual speeds...

Reply:
Also, for me this is more Home Assistant specific, since this function allows us to expose common fan modes such as `LOW`, `MEDIUM` and `HIGH`, but it also adds the rest of the speeds as a percentage. This is done so these speeds can be exposed to Alexa, Homekit...
See https://github.com/home-assistant/core/pull/92840#issuecomment-1542049704